src/torrent.cpp

```diff
@@ -5678,63 +5680,78 @@ namespace {
 
 		boost::scoped_ptr<std::vector<boost::uint8_t> > p(j->buffer.priorities);
 
-		if (j->ret == piece_manager::no_error)
+		if (m_file_priority != *p)
 		{
-			TORRENT_ASSERT(!j->error);
 			m_file_priority = *p;
 			update_piece_priorities();
 		}
-		else
-		{
-			// in this case, some file priorities failed to get set
-			TORRENT_ASSERT(j->error);
-			TORRENT_ASSERT(!m_share_mode);
 
-			if (alerts().should_post<file_error_alert>())
-				alerts().emplace_alert<file_error_alert>(j->error.ec
-					, resolve_filename(j->error.file), j->error.operation_str(), get_handle());
+		if (!j->error) return;
 
-			set_error(j->error.ec, j->error.file);
-			pause();
-		}
+		// in this case, some file priorities failed to get set
+		TORRENT_ASSERT(!m_share_mode);
+
+		if (alerts().should_post<file_error_alert>())
+			alerts().emplace_alert<file_error_alert>(j->error.ec
+				, resolve_filename(j->error.file), j->error.operation_str(), get_handle());
+
+		set_error(j->error.ec, j->error.file);
+		pause();
 	}
 
 	void torrent::prioritize_files(std::vector<int> const& files)
 	{
 		INVARIANT_CHECK;
 
-		// store file priorities until we get the metadata
-		if (!valid_metadata())
-		{
-			m_file_priority.assign(files.begin(), files.end());
-			return;
-		}
+		if (is_seed()) return;
 
 		std::vector<boost::uint8_t> const new_priority = fix_priorities(files, m_torrent_file->files());
 
-		if (!m_torrent_file->num_pieces() || is_seed() || new_priority == m_file_priority) { return; }
```

Comment:
I think a job shouldn't be submitted if file priorities didn't change.

Reply:
but we don't know whether there's an outstanding set_priority job, in flight. i.e. consider the following:

```
h.set_file_priority(0, 0);
h.set_file_priority(0, 4);
```

The second call to `set_file_priority()` will most likely compare equal to the current priorities, because the first one hasn't been applied yet.